# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis with pandas. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [2]:
# Import the appropriate libraries with aliases
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Create two new dataframes
northdf=pd.read_csv("NorthCallCenter.csv")
southdf=pd.read_csv("SouthCallCenter.csv")

## Task 1: Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, answer the following questions. Use the code blocks below to begin cleaning your data. At the end of the section, record your answers.

#### EDA Question 1A:  How many reps are in the North branch?  

In [3]:
# EDA Question 1A solution below:
count_of_north_reps=northdf['Rep ID'].nunique()
print(f"There are", count_of_north_reps, "reps in the North branch.")

There are 9 reps in the North branch.


#### EDA Question 1B:  How many reps are in the South branch?  

In [4]:
# EDA Question 1B solution Below:
count_of_south_reps=southdf['Rep ID'].nunique()
print(f"There are", count_of_south_reps, "reps in the South branch.")

There are 11 reps in the South branch.


#### EDA Question 2A:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the North branch? 

In [5]:
# EDA Question 2A solution Below:
northdf.describe()


Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  245.000000  245.000000  245.000000    245.000000         245.000000   
mean   122.000000    0.020408    0.028571      9.987755           1.914286   
std     70.869599    0.141681    0.166940      1.403980           0.771416   
min      0.000000    0.000000    0.000000      8.000000           1.000000   
25%     61.000000    0.000000    0.000000      9.000000           1.000000   
50%    122.000000    0.000000    0.000000     10.000000           2.000000   
75%    183.000000    0.000000    0.000000     11.000000           3.000000   
max    244.000000    1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count           82.000000             245.000000  245.0  
mean             3.048780               2.967347    1.0  
std              1.456382               0.843827    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

#### EDA Question 2B:  What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for the South branch? 

In [6]:
# EDA Question 2B solution Below:
southdf.describe()

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  314.000000  314.000000  314.000000     314.00000         314.000000   
mean   156.500000    0.035032    0.012739      10.05414           1.910828   
std     90.788215    0.184154    0.112324       1.40751           0.860016   
min      0.000000    0.000000    0.000000       8.00000           1.000000   
25%     78.250000    0.000000    0.000000       9.00000           1.000000   
50%    156.500000    0.000000    0.000000      10.00000           2.000000   
75%    234.750000    0.000000    0.000000      11.00000           3.000000   
max    313.000000    1.000000    1.000000      12.00000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          126.000000             314.000000  314.0  
mean             3.000000               3.076433    1.0  
std              1.308434               0.819424    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

#### EDA Question 3:  What is the number of calls taken for each time block(Both North and South combined)?

In [18]:
# EDA Question 3 solution Below:
totaldf=pd.merge(southdf,northdf, how="outer")
calls_by_block=totaldf.groupby('Time Block')[['Calls']].sum()
print(calls_by_block)

            Calls
Time Block       
10:00 AM       99
11:00 AM       56
12:00 PM      120
1:00 PM        40
2:00 PM        65
3:00 PM        73
4:00 PM        53
5:00 PM        43
9:00 AM        10


##### Record your answers for the EDA section below:
- 1a) Reps in the North Branch = 9
- 1b) Reps in the South Branch = 11


- 2a) North Branch, (round to the nearest hundredth):
> - Busy Minutes = 9.99
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.05
> - During Call Wait Time = 2.97
- 2b) South Branch, (round to the nearest hundredth):
> - Busy Minutes = 10.05
> - Not Ready Minutes = 1.91
> - Incoming Wait Time = 3.0
> - During Call Wait Time = 3.08
- 3) Total Calls taken for all branches per time block:
> - 9:00AM   = 10
> - 10:00AM  = 99
> - 11:00AM  = 56
> - 12:00PM  = 120
> - 1:00PM   = 40
> - 2:00PM   = 65
> - 3:00PM   = 73
> - 4:00PM   = 53
> - 5:00PM   = 43



## Task 2: Cleaning Your Data 
Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. 

In [ ]:
totaldf.info()
#I used the .info function to check if there are any null values in any columns and ensure each column has the correct data type assigned.
#I see the only column with null values is "Incoming Wait Time".
#This function also helps me make sure each of the columns are necessary for analysis (or I could drop them). There aren't any columns I would want to drop at this time.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             559 non-null    int64  
 1   Branch                 559 non-null    object 
 2   Call Purpose           559 non-null    object 
 3   Time Block             559 non-null    object 
 4   Incoming or Outgoing   559 non-null    object 
 5   Queue                  559 non-null    object 
 6   Rep ID                 559 non-null    object 
 7   Sale                   559 non-null    object 
 8   Lost Call              559 non-null    int64  
 9   Abandoned              559 non-null    int64  
 10  Busy Minutes           559 non-null    int64  
 11  Not Ready Minutes      559 non-null    int64  
 12  Incoming Wait Time     208 non-null    float64
 13  During Call Wait Time  559 non-null    int64  
 14  Calls                  559 non-null    int64  
dtypes: flo

In [ ]:
null_df=totaldf[totaldf['Incoming Wait Time'].isnull()]
null_df.sample(10)
#I created a new data frame with only the rows that have a null value for "Incoming Wait Time".
#Then I used Sample to look at those rows. I'm noticing they're all outgoing calls, which makes sense because those calls would have a null wait time.

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
519         274  South  Product Support    3:00 PM             Outgoing     D   
202         101  North    Sales Support   10:00 AM             Outgoing     B   
393         196  South    Sales Support    2:00 PM             Outgoing     D   
244         122  North    Sales Support    1:00 PM             Outgoing     A   
507         262  South  Product Support    2:00 PM             Outgoing     C   
418         209  North  Product Support    3:00 PM             Outgoing     A   
471         235  South  Product Support   12:00 PM             Outgoing     D   
332         166  North  Product Support   12:00 PM             Outgoing     B   
242         121  North    Sales Support    1:00 PM             Outgoing     B   
460         230  North  Product Support    4:00 PM             Outgoing     B   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
519    Kate   NO          0          0            12                  1   
202  Xander   NO          0          0            10                  1   
393   Helga   NO          0          0             8                  2   
244  Amanda   NO          0          0            10                  2   
507   Susan   NO          0          0            11                  1   
418  Amanda   NO          0          0             8                  2   
471   Helga   NO          0          0             8                  3   
332  Xander   NO          0          0             9                  2   
242  Xander   NO          0          0            11                  2   
460    Todd   NO          0          0            12                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
519                 NaN                      3      1  
202                 NaN                      4      1  
393                 NaN                      4      1  
244                 NaN                      3      1  
507                 NaN                      2      1  
418                 NaN                      2      1  
471                 NaN                      2      1  
332                 NaN                      2      1  
242                 NaN                      2      1  
460                 NaN                      3      1

In [ ]:
totaldf.describe()
#I'm using the .describe function to review the minimums & maximum values for each column, there are no obvious outliers.

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  559.000000  559.000000  559.000000    559.000000         559.000000   
mean   141.379249    0.028623    0.019678     10.025045           1.912343   
std     84.340108    0.166893    0.139016      1.405091           0.821644   
min      0.000000    0.000000    0.000000      8.000000           1.000000   
25%     69.500000    0.000000    0.000000      9.000000           1.000000   
50%    139.000000    0.000000    0.000000     10.000000           2.000000   
75%    209.000000    0.000000    0.000000     11.000000           3.000000   
max    313.000000    1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          208.000000             559.000000  559.0  
mean             3.019231               3.028623    1.0  
std              1.365417               0.831225    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

In [ ]:
totaldf['Rep ID'].value_counts().sort_index()
#I'm checking to see if any Rep IDs have duplicates/misspellings.

Rep ID
Alice     20
Amanda    23
Andy      23
Brent     37
Cam       24
Duke      29
Eric      35
George    29
Helga     31
Joe       26
Josh      26
Karl      28
Kate      29
Lilly     30
Randy     33
Sandy     32
Sharon    25
Susan     26
Todd      29
Xander    24
Name: count, dtype: int64

In [ ]:
totaldf['Call Purpose'].value_counts().sort_index()
#Doing the same with the Call Purpose column.

Call Purpose
Complaint           51
Product Support    310
Sales Support      198
Name: count, dtype: int64

In [ ]:
totaldf['Sale'].value_counts()
#And checking the Sale column. Here I'm seeing there is an erroneous YES.

Sale
NO      542
YES      16
YES       1
Name: count, dtype: int64

In [ ]:
totaldf['Sale']=totaldf['Sale'].str.strip()
totaldf['Sale'].value_counts()
#Since I found an extra YES above, I tried stripping the whitespace from everything in the Sale column. Which worked!


Sale
NO     542
YES     17
Name: count, dtype: int64

In [ ]:
totaldf.drop_duplicates()
#Here I am removing any duplicate rows. Since this output still contains 559 rows, that tells me there were no duplicates.

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0             0  North        Complaint    9:00 AM             Incoming     B   
1             0  South    Sales Support   10:00 AM             Incoming     D   
2             1  North    Sales Support   11:00 AM             Incoming     A   
3             1  South    Sales Support   10:00 AM             Incoming     C   
4             2  North    Sales Support   11:00 AM             Incoming     B   
..          ...    ...              ...        ...                  ...   ...   
554         309  South  Product Support    5:00 PM             Outgoing     D   
555         310  South  Product Support    5:00 PM             Outgoing     C   
556         311  South  Product Support    5:00 PM             Outgoing     C   
557         312  South  Product Support    5:00 PM             Outgoing     C   
558         313  South  Product Support    5:00 PM             Outgoing     D   

     Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0     Brent   NO          0          1             9                  2   
1      Kate   NO          0          0             9                  1   
2       Cam   NO          0          0            11                  1   
3      Eric   NO          0          0             8                  2   
4      Todd   NO          0          0             9                  3   
..      ...  ...        ...        ...           ...                ...   
554   Helga   NO          0          0            10                  3   
555   Susan   NO          0          0            12                  3   
556   Sandy   NO          0          0             9                  3   
557    Eric   NO          0          0            10                  1   
558  George   NO          0          0            10                  2   

     Incoming Wait Time  During Call Wait Time  Calls  
0                   1.0                      4      1  
1                   1.0                      2      1  
2                   1.0                      3      1  
3                   1.0                      4      1  
4                   1.0                      3      1  
..                  ...                    ...    ...  
554                 NaN                      4      1  
555                 NaN                      4      1  
556                 NaN                      4      1  
557                 NaN                      2      1  
558                 NaN                      3      1  

[559 rows x 15 columns]

#### Cleaning Question 1:  Is there any missing data and if yes explain what you would do with the missing data and why?

In [4]:
# Question 1 solution below

#### Record your answer below:

> Your Answer: I only found null data in the "Incoming Call Wait Time" column. I chose not to edit the data in that column, because it's accurate the data is null. The rows with null values are because they are outgoing calls--which wouldn't have a wait time (0 would be inaccurate).

#### Cleaning Question 2:  In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?

In [8]:
# Cleaning Question 2 solution below:

##### Record your answer by below:
> Your Answer: There were 2 YES values, which I suspected was because of a leading or trailing whitespace. I used the strip function to eliminate whitespaces from all values in that column, which did remove the separate YES.

#### Cleaning Question 3:  Are there duplicates in the two data sets? If there are, how did you handle them?

In [9]:
# Cleaning Question 3 solution below:

##### Record your answer below:
> Your Answer: I did not find any duplicates after using the .drop_duplicates function.

#### Cleaning Question 4:  Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [17]:
# Cleaning Question 4 solution below:

##### Record your answer below:
> Your Answer: I am choosing to believe there is no unnecessary data in the set at this time because the questions we're looking at seem relevant to all columns (at least for the exploration stage).

## Task 3: Data Manipulation
Before you begin answering any questions, combine the two datasets together to create a third dataframe called df. You can use this third dataframe to compare the two call centers to company-wide trends.
* Add additional code blocks as needed to show how you came to your conclusions. Add comments in your code blocks to help others understand your thinking. Record your answer below.

In [46]:
# Create dataframe for the entire company named df
df=totaldf
#I did this earlier, whoops.


#### Manipulation Question 1:  Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.

In [12]:
# Manipulation Question solution below:

##### Record your answer below
Rep with the hightest number of calls and their total calls:
- North Branch = 
- South Branch = 

#### Manipulation Question 2:  The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?

In [13]:
# Manipulation Question 2 solution below:

##### Record your answer below:
Average Call Abandonment Rates (round to the nearest hundredth):
- North Branch = 
- South Branch = 
- Company Wide =
- Do any of these fall out of the optimal range of 2-5%?

#### Manipulation Question 3:  Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?

In [14]:
# Manipulation Question 3 solution below:

##### Record your answer below:
Percentage of calls answered within 2 seconds, include # of calls:
- North Branch = 
- South Branch = 
- Company Wide = 

#### Manipulation Question 4: For each branch and the entire company, what is the average speed of answer?

In [15]:
# Manipulation Question 4 solution below:

##### Record your answer by below:
Average speed of answer (rounded to nearest hundredth):
- North Branch in seconds = 
- South Branch in seconds = 
- Company Wide in seconds = 

## Task 4: Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why. 

*NOTE Some questions you may decide to use more than one chart and or chart style.

#### Visualization 1:  What is the average abandonment rate per queue?

In [1]:
# Create visualization 1 here
# Import any additional libraries needed with alias

# The dictionary abandonment_rates has the data you need.
abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

#Your code below:


NameError: name 'df' is not defined

##### Record your answer below:

> Chart style you chose and why: 

#### Visualization 2: What is the service level percentage and average speed of answer for each rep in the North Branch?

In [2]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

#Your Code Here:


NameError: name 'north' is not defined

##### Record your answer below:

> Chart style you chose and why: 

#### Visualization 3: For each type of call purpose, how many calls are outgoing vs. incoming?

In [3]:
# Create visualization 3 here:
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_group = df.groupby("Call Purpose")
call_purpose = ["Complaint", "Product Support", "Sales Support"]
purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
print(purpose_counts)

complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

#Your Code Here:


NameError: name 'df' is not defined

##### Record your answer below:

> Chart style you chose and why: 

## Task 5: Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.


#### Summary Question 1:  Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?

##### Record your answer below:
> Your Answer: 



#### Summary Question 2: Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?

##### Record your answer below:
> Your Answer: 


#### Summary Question 3: Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

##### Record your answer below:
> Your Answer: 


## Bonus Mission
Create a visualization that answers this question: For each call purpose, how many calls (incoming and outgoing) take place in each time block?
##### Record your answer below:

> Chart style you chose and why: 

In [47]:
# Create your Bonus Mission visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot: